### PR TITLE
[merged] Options {DOCKER_ROOT_VOLUME, DOCKER_ROOT_VOLUME_SIZE} for setting up separate logical volume for docker root.

### PR DESCRIPTION
**Pull request by shishir-a412ed**
Signed-off-by: Shishir Mahajan <shishir.mahajan@redhat.com>

### CODE REVIEW

**Comment by shishir-a412ed** [comment]
ping @rhvgoyal

File: docker-storage-setup.sh [hunk 1]
```diff
@@ -873,6 +987,82 @@ setup_storage() {
...
+    return 1
+  fi
+}
+
+# Inherit the same selinux context as /var/lib/docker.
+setup_selinux_context_root_dir() {
+  local dir=$1
+  local policy
+
+  [ "$dir" == "/var/lib/docker" ] && return 0
```
Comment (rhatdan) [inline docker-storage-setup.sh]: Shouldn't this be ${DOCKER_ROOT_DIR}

File: docker-storage-setup.sh [hunk 2]
```diff
@@ -873,6 +987,82 @@ setup_storage() {
...
+
+  # If there is already a labeling policy for the dir, don't try to add one.
+  policy=`semanage fcontext -l -C $dir | grep "^$dir =" | wc -l`
```
Comment (rhatdan) [inline docker-storage-setup.sh]: Shouldn't this be?

```
policy=`semanage fcontext -l -C | grep "^$dir =" | wc -l`
```

File: docker-storage-setup.sh [hunk 3]
```diff
@@ -873,6 +987,82 @@ setup_storage() {
...
+    if ! semanage fcontext -a -e /var/lib/docker $dir; then
+      echo "semanage fcontext -a -e /var/lib/docker $dir failed."
+      return 1
+    fi
+  fi
+}
+
+get_docker_root_dir(){
+    local flag=false
+    options=$(grep -e "^OPTIONS" /etc/sysconfig/docker|cut -d"'" -f 2)
```
Comment (rhatdan) [inline docker-storage-setup.sh]: How about something like this

 sh -c ". /etc/sysconfig/docker; echo \$OPTIONS"
--selinux-enabled --log-driver=journald --add-registry atomic-registry.usersys.redhat.com

File: docker-storage-setup.sh [hunk 4]
```diff
@@ -873,6 +987,82 @@ setup_storage() {
...
+    for opt in $options
```
Comment (rhatdan) [inline docker-storage-setup.sh]: I think we can match the regular expression using something like awk or grep rather then building our own version.

Comment (rhatdan) [inline docker-storage-setup.sh]: Do we handle the case where we have docker-latest installed?

**Comment by rhvgoyal** [comment]
Can we provide better changelog for the patch and explain what problem these new options are solving.

File: docker-storage-setup.1 [hunk 5]
```diff
@@ -63,6 +63,21 @@ GROWPART:
       disabled. Use GROWPART=true to enable automatic partition
       table resizing.
 
+DOCKER_ROOT_VOLUME:
+      If set to "yes", then a separate logical volume will be setup
+      for docker root and it will be mounted at docker root directory.
+      By default, docker root is located at "/var/lib/docker", unless a
+      user start docker daemon using "-g" or "--graph" option and provides
+      a custom docker root location.  
+
```
Comment (rhvgoyal) [inline docker-storage-setup.1]: Probably it is a good idea to mention that we are going to scan /etc/sysconfig/docker
file to figure out -g option. If found it will be used as docker root dir otherwise /var/lib/docker is default.

File: docker-storage-setup.conf [hunk 6]
```diff
@@ -1,7 +1,28 @@
...
+# values # acceptable to "lvcreate -l", then only those values which
+# contains "%" in syntax are acceptable.  If value does not contain "%" it
+# is assumed # value is suitable for "lvcreate -L".
+DOCKER_ROOT_VOLUME_SIZE=40%FREE
+
 # Specify storage driver one wants to use with docker. Default is devicemapper.
 # Other possible options are overlay, overlay2 and "". Empty string means do
 # not do any storage setup.
-#
+
```
Comment (rhvgoyal) [inline docker-storage-setup.conf]: No need for this extra line.

File: docker-storage-setup.conf [hunk 7]
```diff
@@ -1,7 +1,28 @@
...
+# The desired size for the docker root volume. It defaults using 40% of all
+# free space.
+#
+# DOCKER_ROOT_VOLUME_SIZE can take values acceptable to "lvcreate -L" as well
+# as some values acceptable to to "lvcreate -l". If user intends to pass
+# values # acceptable to "lvcreate -l", then only those values which
+# contains "%" in syntax are acceptable.  If value does not contain "%" it
+# is assumed # value is suitable for "lvcreate -L".
+DOCKER_ROOT_VOLUME_SIZE=40%FREE
+
```
Comment (rhvgoyal) [inline docker-storage-setup.conf]: These are new options. Instead of adding them in the beginning of file, how about adding them towards the end of the file?

File: docker-storage-setup.conf [hunk 8]
```diff
@@ -1,7 +1,28 @@
+# By default no new volume and fs will be setup for docker root dir. Docker
+# will create /var/lib/docker/ and use underlying fs for storing image and
+# containers.
+# 
+# If one wants to carve out a separate logical volume for storing docker
+# metadata and images/containers/volumes, then set DOCKER_ROOT_VOLUME=yes
+#
+# This will try to create a logical volume, create filesystem on that volume
+# and mount on $DOCKER_ROOT_DIR.
```
Comment (rhvgoyal) [inline docker-storage-setup.conf]: I think very soon we will have request for allowing passing in filesystem to create and options to pass in. May be we can create two more options at that time. Say,

DOCKER_ROOT_FS="xfs"
DOCKER_ROOT_FS_OPTS="-o discard"

This is just an example for future extension. For now, I think we can just mention that by default xfs filesystem is created.

File: docker-storage-setup.conf [hunk 9]
```diff
@@ -1,7 +1,28 @@
...
+DOCKER_ROOT_VOLUME=yes
+
```
Comment (rhvgoyal) [inline docker-storage-setup.conf]: By default we don't want to create docker root volume. So this should be.

DOCKER_ROOT_VOLUME=no

File: docker-storage-setup.conf [hunk 10]
```diff
@@ -1,7 +1,28 @@
...
+# The desired size for the docker root volume. It defaults using 40% of all
+# free space.
+#
+# DOCKER_ROOT_VOLUME_SIZE can take values acceptable to "lvcreate -L" as well
+# as some values acceptable to to "lvcreate -l". If user intends to pass
+# values # acceptable to "lvcreate -l", then only those values which
```
Comment (rhvgoyal) [inline docker-storage-setup.conf]: Extra `#` in the middle of line.

File: docker-storage-setup.conf [hunk 11]
```diff
@@ -1,7 +1,28 @@
...
+# contains "%" in syntax are acceptable.  If value does not contain "%" it
+# is assumed # value is suitable for "lvcreate -L".
```
Comment (rhvgoyal) [inline docker-storage-setup.conf]: Extra `#` in the middle of line.

File: docker-storage-setup.sh (hunk 4 above)
Comment (shishir-a412ed) [inline docker-storage-setup.sh]: Fixed based on nalin's comments.

Comment (shishir-a412ed) [inline docker-storage-setup.sh]: docker-latest will be handled in the spec file, not part of this PR.

File: docker-storage-setup.sh (hunk 3 above)
Comment (shishir-a412ed) [inline docker-storage-setup.sh]: @rhvgoyal is okay with `grep`. Leaving as it is for now.

**Comment by shishir-a412ed** [comment]
@rhvgoyal @rhatdan PTAL.

**Comment by rhvgoyal** [comment]
shishir, lets keep changelog line lenth to around 60-70. Right now, do git log on your terminal and chanelog lines are much longer and these are wrapping and look pretty odd.

File: docker-storage-setup.conf [hunk 12]
```diff
@@ -84,3 +83,24 @@ DEVICE_WAIT_TIMEOUT=60
 # by setting WIPE_SIGNATURES=true. Be careful before using this option
 # as this means if there was any leftover data on disk, it will be lost.
 WIPE_SIGNATURES=false
+
+# By default no new volume and filesystem (fs) will be setup for docker root dir.
+# Docker will create /var/lib/docker/ and use underlying fs for storing image and
```
Comment (rhvgoyal) [inline docker-storage-setup.conf]: line more than 80 char.

File: docker-storage-setup.conf [hunk 13]
```diff
@@ -84,3 +83,24 @@ DEVICE_WAIT_TIMEOUT=60
 # by setting WIPE_SIGNATURES=true. Be careful before using this option
 # as this means if there was any leftover data on disk, it will be lost.
 WIPE_SIGNATURES=false
+
+# By default no new volume and filesystem (fs) will be setup for docker root dir.
```
Comment (rhvgoyal) [inline docker-storage-setup.conf]: line more than 80 char

File: docker-storage-setup.sh [hunk 14]
```diff
@@ -865,6 +969,16 @@ setup_storage() {
    Fatal "Storage is already configured with ${current_driver} driver. Can't configure it with ${STORAGE_DRIVER} driver. To override, remove $DOCKER_STORAGE and retry."
   fi
 
+  if [ "$STORAGE_DRIVER" == "devicemapper" ] && [ "$DOCKER_ROOT_VOLUME" == "yes" ];then
+   Fatal "STORAGE_DRIVER=${STORAGE_DRIVER} and DOCKER_ROOT_VOLUME=${DOCKER_ROOT_VOLUME} are mutually exclusive options. Use either one of them, not both."
+  fi
```
Comment (rhvgoyal) [inline docker-storage-setup.sh]: shishir, I am changing my mind on this now. There is no real technical reason that why root volume can't be allowed with devicemapper storage driver. So lets get rid of this restriction.

File: docker-storage-setup.sh [hunk 15]
```diff
@@ -865,6 +969,16 @@ setup_storage() {
...
+
+  # If docker root is on a separate volume, setup that first.
+  if ! setup_docker_root_lv_fs; then
+    echo "Failed to setup docker root volume."
+    return 1
+  fi
+
```
Comment (rhvgoyal) [inline docker-storage-setup.sh]: I guess we will have to define in documentation very clearly order of creation of root lv and devicemapper lv. Reason being that they both can pass syntax like 50%FREE. And order of operation will decide which gets what size.

To make sure we don't get surprises, how about creating devicemapper volume first and then root volume.

Comment (rhvgoyal) [inline docker-storage-setup.sh]: Also lets document this in man page and also in docker-storage-setup.conf

**Comment by shishir-a412ed** [comment]
@rhvgoyal PTAL.

**Comment by rhvgoyal** [comment]
When i do `man ./docker-storage-setup.1`, description of DOCKER_ROOT_VOLUME and DOCKER_ROOT_VOLUME_SIZE is not properly formatted. Some of the lines are too long.

**Comment by rhvgoyal** [comment]
In man page, can we add description of new options after `WIPE_SIGNATURES`. This will be consistent with docker-storage-setup.conf

File: docker-storage-setup.sh (hunk 3 above)
Comment (rhvgoyal) [inline docker-storage-setup.sh]: Can we define path to docker file in a global variable instead of hardcoding `/etc/sysconfig/docker` here.

Comment (rhvgoyal) [inline docker-storage-setup.sh]: I am wondering if we can use /usr/bin/getopt utility to parse -g/--graph-driver parameter? An example is provided in `/usr/share/doc/util-linux/getopt-parse.bash`

**Comment by rhvgoyal** [comment]
We will have to change "reset_storage" as well I think to remove docker root volume. Also we probably will have to unmount volume first and then remove it.

**Comment by rhvgoyal** [comment]
Can we add some automated test cases to test this feature.

File: docker-storage-setup.sh [hunk 16]
```diff
@@ -865,14 +969,102 @@ setup_storage() {
...
+    options=$(grep -e "^OPTIONS" /etc/sysconfig/docker|cut -d"'" -f 2)
+    for opt in $options
+    do
+        if [ "$flag" = true ];then
+           DOCKER_ROOT_DIR=$opt
+           flag=false
+           continue
+        fi
+        if [[ "$opt" == --graph* || "$opt" == -g* ]];then
+           if [[ "$opt" == *"="* ]];then
```
Comment (rhvgoyal) [inline docker-storage-setup.sh]: This also accepts `--graph-driver foo` as valid option.

File: docker-storage-setup.sh [hunk 17]
```diff
@@ -865,14 +969,102 @@ setup_storage() {
...
+           else
+             flag=true
+             continue
+           fi
+        fi
+    done
+}
+
+check_docker_root_volume_options() {
+
```
Comment (rhvgoyal) [inline docker-storage-setup.sh]: This function name does not seem to match what we are doing in that function. Basically we are parsing and setting up docker root dir. May be call this function `setup_docker_root_dir`

File: docker-storage-setup.sh [hunk 18]
```diff
@@ -839,6 +841,108 @@ get_existing_storage_driver() {
...
+    fi
+
+    if ! mount_docker_root_volume; then
+      return 1
+    fi
+
+    # setup right selinux label first time fs is created. Mount operation
+    # changes the label of directory to reflect the label on root inode
+    # of mounted fs.
+    if ! restore_selinux_context $DOCKER_ROOT_DIR; then
```
Comment (rhatdan) [inline docker-storage-setup.sh]: You could do a mkdir -Z when creating the directory, But this will not create it with the correct label if the directory is /host/var/lib/docker.  I just submitted some changes to container-selinux that would make the creation of this directory work for certain process types.  We probably can get it to work correctly for docker-storage-setup, if we know which domain it is going to run as.  If it runs as docker_t these directories will get created with the correct label,

File: docker-storage-setup.sh (hunk 16 above)
Comment (shishir-a412ed) [inline docker-storage-setup.sh]: Which `docker version` do you see this flag (`--graph-driver`). I am not able to find it.

File: docker-storage-setup.sh (hunk 18 above)
Comment (rhvgoyal) [inline docker-storage-setup.sh]: @rhatdan, so you control the label of docker-storage-setup in selinux policy right? So you can control that to whatever you want and make sure directory is created with right label?

File: docker-storage-setup.sh (hunk 16 above)
Comment (rhvgoyal) [inline docker-storage-setup.sh]: This is not a valid option. I just set invalid option --graph-driver in /etc/sysconfig/docker and this logic still took it.

**Comment by shishir-a412ed** [comment]
@rhvgoyal @rhatdan PTAL.

Shishir

File: docker-storage-setup.conf [hunk 19]
```diff
@@ -84,3 +83,28 @@ DEVICE_WAIT_TIMEOUT=60
...
 WIPE_SIGNATURES=false
+
+# By default no new volume and filesystem (fs) will be setup for docker
+# root dir. Docker will create /var/lib/docker/ and use underlying fs for
+# storing image and containers.
+#
+# If one wants to carve out a separate logical volume for storing docker
+# metadata and images/containers/volumes, then set DOCKER_ROOT_VOLUME=yes
+#
+# This will try to create a logical volume, create a XFS filesystem on
```
Comment (rhatdan) [inline docker-storage-setup.conf]: docker-storage-setup will try to create ...

File: docker-storage-setup.conf [hunk 20]
```diff
@@ -84,3 +83,28 @@ DEVICE_WAIT_TIMEOUT=60
...
+# that volume and mount on $DOCKER_ROOT_DIR.
+#
+# NOTE: If both STORAGE_DRIVER=devicemapper and DOCKER_ROOT_VOLUME=yes is
+# passed, docker-storage-setup would set up the thin pool for devicemapper
```
Comment (rhatdan) [inline docker-storage-setup.conf]: s/passed/set

File: docker-storage-setup.conf [hunk 21]
```diff
@@ -84,3 +83,28 @@ DEVICE_WAIT_TIMEOUT=60
...
+# first, followed by docker root volume.
+DOCKER_ROOT_VOLUME=no
+
+# The desired size for the docker root volume. It defaults using 40% of all
```
Comment (rhatdan) [inline docker-storage-setup.conf]: Shouldn't this start out with a 
DOCKER_ROOT_VOLUME_SIZE

Is this the Percentage of free space available 
The default size for DOCKER_ROOT_VOLUME_SIZE

Comment (rhatdan) [inline docker-storage-setup.conf]: I think this whole description section we should work on together, if this is the section an admin would be reading.

Comment (rhatdan) [inline docker-storage-setup.conf]: We can probably merge the this patch and then sit down in a room and go over the description line by line,

File: docker-storage-setup.sh [hunk 22]
```diff
@@ -839,6 +853,141 @@ get_existing_storage_driver() {
...
+  [ -z "$mounts" ] && return 1
+  # If root volume is mounted, it should be mounted on $DOCKER_ROOT_DIR
+  # otherwise it is an error.
+  target=`echo $mounts | cut -d " " -f1`
+  [ "$target" == "$DOCKER_ROOT_DIR" ] && return 0
+    echo "Docker root volume /dev/$VG/$DOCKER_ROOT_LV_NAME is mounted on $target instead of ${DOCKER_ROOT_DIR}. Unmount it and try again."
+    exit 1
+}
+
+# drop a systemd mount file at /etc/systemd/system.
```
Comment (rhatdan) [inline docker-storage-setup.sh]: Mount unit configuration file

File: docker-storage-setup.sh [hunk 23]
```diff
@@ -839,6 +853,141 @@ get_existing_storage_driver() {
...
+# This would allow the $DOCKER_ROOT_DIR mountpoint to persist across reboots.
+drop_systemd_mount_file() {
+  IFS='/ '
+  # filename must match $DOCKER_ROOT_DIR path. e.g if $DOCKER_ROOT_DIR is /var/lib/docker
+  # then filename will be var-lib-docker.mount
+  read -r -a elements <<< $DOCKER_ROOT_DIR
```
Comment (rhatdan) [inline docker-storage-setup.sh]: filename=echo ${DOCKER_ROOT_DIR}.mount /var/lib/docker | sed 's|/|-|g' | cut -b 2-

File: docker-storage-setup.sh [hunk 24]
```diff
@@ -839,6 +853,141 @@ get_existing_storage_driver() {
...
+  [ "$target" == "$DOCKER_ROOT_DIR" ] && return 0
+    echo "Docker root volume /dev/$VG/$DOCKER_ROOT_LV_NAME is mounted on $target instead of ${DOCKER_ROOT_DIR}. Unmount it and try again."
+    exit 1
+}
+
+# drop a systemd mount file at /etc/systemd/system.
+# This would allow the $DOCKER_ROOT_DIR mountpoint to persist across reboots.
+drop_systemd_mount_file() {
+  IFS='/ '
+  # filename must match $DOCKER_ROOT_DIR path. e.g if $DOCKER_ROOT_DIR is /var/lib/docker
```
Comment (rhvgoyal) [inline docker-storage-setup.sh]: If you are setting IFS, then you should save previous IFS and restore it at the end?

**Comment by shishir-a412ed** [comment]
@rhatdan @rhvgoyal PTAL.

Shishir

**Comment by rhatdan** [comment]
@shishir-a412ed Lets meet with @rhvgoyal and go over the docs before we merge.  Meet after standup?

**Comment by shishir-a412ed** [comment]
@rhvgoyal @rhatdan PTAL.

One last thing I wanted to check: Should we cleanup the mount file on `/usr/bin/docker-storage-setup --reset`? i.e in `reset_docker_root_volume` function, should we do:
```
systemctl disable $filename`
systemctl daemon-reload
rm /etc/systemd/system/$filename
```
Shishir

**Comment by rhvgoyal** [comment]
@shishir-a412ed I think cleaning up systemd .mount file on reset makes sense. Anyway we will delete the root logical volume and mounting will fail. So yes, lets clean it up.

**Comment by rhatdan** [comment]
Make sure you eat errors of $filename is not enabled.

File: docker-storage-setup.1 [hunk 25]
```diff
@@ -93,11 +93,39 @@ DEVICE_WAIT_TIMEOUT:
 WIPE_SIGNATURES:
       Wipe any signatures found on disk. Valid values are
       true/false and default value is false. By default if any
-      signatures are found on disk operation is aborted. If this value
-      is set to true, then signatures will either be wiped or
+      signatures are found on disk operation is aborted. If this
+      value is set to true, then signatures will either be wiped or
```
Comment (rhvgoyal) [inline docker-storage-setup.1]: Why are we changing this? This is not related to this patch.

File: docker-storage-setup.1 [hunk 26]
```diff
@@ -93,11 +93,39 @@ DEVICE_WAIT_TIMEOUT:
...
+      By default no new volume and filesystem will be setup for docker
+      root dir. Docker creates /var/lib/docker/ on top of underlying
+      filesystem for storing images and containers.
+
+      To carve out a separate logical volume for storing docker
+      images/containers/volumes data, set DOCKER_ROOT_VOLUME=yes
+      NOTE: devicemapper stores images/containers on thin pool.
+
+      docker-storage-setup creates a logical volume with an XFS
+      filesystem mounted on docker root directory (default: /var/lib/docker).
```
Comment (rhvgoyal) [inline docker-storage-setup.1]: Do man `./docker-storage-setup.1` and make sure all the lines look fine. This line seems to wrap.

File: docker-storage-setup.1 [hunk 27]
```diff
@@ -93,11 +93,39 @@ DEVICE_WAIT_TIMEOUT:
...
+
+DOCKER_ROOT_VOLUME_SIZE:
+      Specify the desired size for docker root volume. It defaults to
+      40% of all free space.
+
+      DOCKER_ROOT_VOLUME_SIZE can take values acceptable to "lvcreate -L"
```
Comment (rhvgoyal) [inline docker-storage-setup.1]: Many of the lines in this section are wrapping.

File: docker-storage-setup.sh [hunk 28]
```diff
@@ -410,6 +412,18 @@ check_docker_storage_metadata() {
   exit 1
 }
 
+reset_docker_root_volume () {
+  if docker_root_volume_exists; then
+     if ! get_docker_root_dir; then
+        return 1
+     fi
```
Comment (rhvgoyal) [inline docker-storage-setup.sh]: We should not have to look at docker root dir again. We can just check if root volume is mounted and where it is mounted. Just unmount it and delete volume.

File: docker-storage-setup.sh [hunk 29]
```diff
@@ -839,6 +853,123 @@ get_existing_storage_driver() {
...
+    lvs $VG/$DOCKER_ROOT_LV_NAME > /dev/null 2>&1 && return 0
+    return 1
+}
+
+# Check if docker root volume is already mounted.
+# This return 0 if volume is already mounted on $DOCKER_ROOT_DIR. Returns 1
+# if volumes needs mounting. On all other error conditions it exits.
+
+docker_root_lv_mounted() {
+  local mounts target
```
Comment (rhvgoyal) [inline docker-storage-setup.sh]: I think we can modify this function to return location of mount point of docker volume. It will return 1 in case of error. Return 0 in case of success or if volume is not mounted. If volume is not mounted, will return 0 and echo empty string.

Comment (rhvgoyal) [inline docker-storage-setup.sh]: That way it can easily be used in more than one context.

File: docker-storage-setup.sh [hunk 30]
```diff
@@ -839,6 +853,123 @@ get_existing_storage_driver() {
...
+  if ! mounts=$(findmnt -n --first-only --source /dev/$VG/$DOCKER_ROOT_LV_NAME); then
+    return 1
+  fi
+  [ -z "$mounts" ] && return 1
+  # If root volume is mounted, it should be mounted on $DOCKER_ROOT_DIR
+  # otherwise it is an error.
+  target=`echo $mounts | cut -d " " -f1`
+  [ "$target" == "$DOCKER_ROOT_DIR" ] && return 0
+    Error "Docker root volume /dev/$VG/$DOCKER_ROOT_LV_NAME is mounted on $target instead of ${DOCKER_ROOT_DIR}. Unmount it and try again."
+    exit 1
```
Comment (rhvgoyal) [inline docker-storage-setup.sh]: This logic of giving error if volume is not mounted on correct directory can go outside the funciton at the call site.

Comment (rhvgoyal) [inline docker-storage-setup.sh]: If you want to exit immediately, you can use `Fatal` instead of `Error`

File: docker-storage-setup.sh [hunk 31]
```diff
@@ -839,6 +853,123 @@ get_existing_storage_driver() {
...
+  Description=Mount docker-root-lv on docker root directory.
+
+  [Mount]
+  What=/dev/$VG/$DOCKER_ROOT_LV_NAME
+  Where=${DOCKER_ROOT_DIR}
+  Type=xfs
+  Options=defaults
+
+  [Install]
+  WantedBy=multi-user.target
```
Comment (rhvgoyal) [inline docker-storage-setup.sh]: I did `systemd list-dependencies --all docker.service` and it does not list `multi-user.target`. That means there is no guarantee that this mount unit will finish execution before docker starts.

How about doing WantedBy=docker.service and that should ensure that this activates

Comment (rhvgoyal) [inline docker-storage-setup.sh]: I am thinking this probably should be `RequiredBy` dependency. That is if we fail to mount this device, then docker should fail.

Comment (rhatdan) [inline docker-storage-setup.sh]: I would prefer not to tie this script to docker if we can help it.  I want to start looking at converting the script to container-storage-setup and being able to use it with OCID.

I am pretty sure that the mount unit will have to be mounted before Multi-User Target and Docker.service or OCID.service will not be started until multi-user target is started.  No need to link them.

Comment (rhatdan) [inline docker-storage-setup.sh]: If you look at docker.service, you will see that it is 
[Install]
WantedBy=multi-user.target

Which means when you enable the docker,service it gets started when systemd starts the multi-user-target.

Comment (rhvgoyal) [inline docker-storage-setup.sh]: @rhatdan IIUC, so multi-user.target wants docker.service. And it also wants docker-var-lib.mount. That means systemd can activate docker.service and docker-var-lib.mount in any order. There is no guarantee that docker-var-lib.mount will finish execution before docker.service. (Atleast multi-user.target does not seem to guarantee that)

Comment (rhvgoyal) [inline docker-storage-setup.sh]: Should we use `WantedBy=local-fs.target` instead? And possibly add `Before=local-fs.target` as well. 

man `systemd.special` says systemd-fstab-generator automatically adds dependencies for entries in /etc/fstab file. Given this will not be in fstab file, that probably means we will have to add dependencies ourselves.

Comment (rhatdan) [inline docker-storage-setup.sh]: Yes I was thinking the same thing. Should it be INSTALL local-fs.target?

**Comment by rhatdan** [comment]
Systemd guys say to add

 Before=local-fs.target

File: docker-storage-setup.sh (hunk 3 above)
Comment (rhvgoyal) [inline docker-storage-setup.sh]: @shishir-a412ed my comment of using a variable for `/etc/sysconfig/docker` is yet to be addressed.

File: docker-storage-setup.sh [hunk 32]
```diff
@@ -865,15 +990,74 @@ setup_storage() {
...
+    do
+        if [ "$flag" = true ];then
+           DOCKER_ROOT_DIR=$opt
+           flag=false
+           continue
+        fi
+	case "$opt" in
+            "-g"|"--graph")
+                flag=true
+                ;;
```
Comment (rhvgoyal) [inline docker-storage-setup.sh]: Does this handle the case of two or more spaces -g and value. For example `-g    /var/lib/foo`

File: docker-storage-setup.sh [hunk 33]
```diff
@@ -410,6 +412,19 @@ check_docker_storage_metadata() {
   exit 1
 }
 
+reset_docker_root_volume () {
+  local mp
+  if docker_root_volume_exists; then
+     if ! mp=$(docker_root_lv_mountpoint); then
+        return 1
```
Comment (rhvgoyal) [inline docker-storage-setup.sh]: It will be a good idea to give an error message here. Say

`Failed to determine where /dev/$VG/$DOCKER_ROOT_LV is mounted`

File: docker-storage-setup.sh [hunk 34]
```diff
@@ -839,6 +854,103 @@ get_existing_storage_driver() {
...
 }
 
+docker_root_volume_exists() {
+    lvs $VG/$DOCKER_ROOT_LV_NAME > /dev/null 2>&1 && return 0
+    return 1
+}
+
+# This returns the mountpoint of $DOCKER_ROOT_LV_NAME
+docker_root_lv_mountpoint() {
+  local mounts target
```
Comment (rhvgoyal) [inline docker-storage-setup.sh]: local variable `target` is unused now.

File: docker-storage-setup.sh (hunk 31 above)
Comment (rhvgoyal) [inline docker-storage-setup.sh]: @shishir-a412ed Change this file to add "Before=local-fs.target" in [Unit] section. And I think we can retain WantedBy=local-fs.target in [Install] section. Though I am not sure if WantedBy can go in [Unit] section or not.

**Comment by rhvgoyal** [comment]
@shishir-a412ed I am testing this patch now.

After creating docker-root-lv, I tried "atomic storage reset" and it asks me whether I really want to remove an active volume. I don't think we need it.  Like lvm thin pool, lets first deactivate the docker root volume and then remove it to get rid of this message.

# atomic storage reset
Do you really want to remove active logical volume fedora/docker-root-lv? [y/n]: y
Removed /etc/systemd/system/local-fs.target.wants/var-lib-docker.mount.

Also I see this "Removed var-lib-docker.mount" message. We don't need to see it. I would say send stdout to /dev/null/

**Comment by rhvgoyal** [comment]
After doing "atomic storage reset" once, I did it again and got following.

atomic storage reset
Failed to disable unit: No such file or directory

We should not have assumption that unit file will be there necessarily.

**Comment by rhvgoyal** [comment]
Also, I unmounted /var/lib/docker manually and did "atomic storage reset" and it complains failed to determine mount point. That should not happen. we should not enforce that file is mounted.

atomic storage reset
ERROR: Failed to determine where /dev/fedora/docker-root-lv is mounted

**Comment by rhvgoyal** [comment]
@rhatdan I removed device from my VM and rebooted. System waits for 1 minute 30 seconds for device to come up. Till then boot stops. 

After 1 minute 30 second timeout, boot continues.

**Comment by rhvgoyal** [comment]
Well, another problem is that docker starts and sets up new overlay2 on /var/lib/docker/. So if disk shows up during next reboot, then docker will switch to old view again.

And all this is happening because docker does not fail if docker-storage-setup fails. This automatic creation of loop devices need is creating lot of problems.

**Comment by rhatdan** [comment]
Is docker-storage-setup a systemd unit file?  If yes put the requirement there?  That the mount has to succeed.

**Comment by rhvgoyal** [comment]
Problem is not docker-storage-setup but docker. Docker unit file has Wants= dependency on docker-storage-setup and not Requires= dependency. That means docker will continue to start even if docker-storage-setup failed.

**Comment by rhvgoyal** [comment]
@shishir-a412ed had a conversation with dan and here is the summary.

- Change dependency in mount unit file to docker-storage-setup instead of local-fs.target

  Before=docker-storage-setup.service
  WantedBy=docker-storage-setup.service

  This should help with issue of whole boot hanging for 90 seconds if device or volume group or logical volume does not show up.

- We will soon have default of overlay2 in fedora. At that time we will modify docker.service file to Requires=docker-storage-setup as well. That will make sure if docker-storage-setup fails, then
docker fails as well. That way we will not run into odd situations if underlying disk or volume group does not come up.

**Comment by rhvgoyal** [comment]
@shishir-a412ed I tested it with `-g foo` in /etc/sysconfig/docker and it gives following error.


./docker-storage-setup.sh 
INFO: Device /dev/vdd is already partitioned and is part of volume group docker-vg
  Wiping xfs signature on /dev/docker-vg/docker-root-lv.
  Logical volume "docker-root-lv" created.
Job for oo.mount failed.
See "systemctl status oo.mount" and "journalctl -xe" for details.
ERROR: Failed to mount docker root volume docker-root-lv on foo

File: docker-storage-setup.sh [hunk 35]
```diff
@@ -410,6 +412,36 @@ check_docker_storage_metadata() {
...
+  fi
+  # If the user has manually unmounted ${DOCKER_ROOT_DIR}, mountpoint (mp)
+  # will be empty. Extract ${mp} from /etc/sysconfig/docker in that case.
+  if [ -z "$mp" ];then
+    if ! get_docker_root_dir; then
+      return 1
+    fi
+    mp=${DOCKER_ROOT_DIR}
+  fi
+  filename=$(echo ${mp}.mount|sed 's|/|-|g' | cut -b 2-)
```
Comment (rhvgoyal) [inline docker-storage-setup.sh]: I think here you need to make sure that mp starts with /.

**Comment by rhvgoyal** [comment]
I tried passing `-g /var/lib/foo/` and that fails too.

**Comment by shishir-a412ed** [comment]
@rhvgoyal I tried this "I tried passing -g /var/lib/foo/ and that fails too."
For me, d-s-s is able to handle this case, remove the docker root volume and exit gracefully.

I will come to your desk, and we can check this out.

Shishir

**Comment by rhvgoyal** [comment]
./docker-storage-setup.sh 
INFO: Device /dev/vdd is already partitioned and is part of volume group docker-vg
  Wiping xfs signature on /dev/docker-vg/docker-root-lv.
  Logical volume "docker-root-lv" created.
Failed to start var-lib-foo-.mount: Unit var-lib-foo-.mount is not loaded properly: Invalid argument.
See system logs and 'systemctl status var-lib-foo-.mount' for details.
ERROR: Failed to mount docker root volume docker-root-lv on /var/lib/foo/

**Comment by rhvgoyal** [comment]
Lets use `realpath -m <path>` to canonicalize path and then use sed on this.

**Comment by rhvgoyal** [comment]
LGTM

I have done lot of local testing and it seems to work fine for me. Once jenkins is working again, I will merge this PR.

**Comment by rhvgoyal** [comment]
Once this gets merged we need two things

- some automated test to test this.
- facility to save local metadata so that we can save filename of mount unit file.

**Comment by rhvgoyal** [comment]
@rh-atomic-bot retest

**Comment by rhvgoyal** [comment]
@rh-atomic-bot retest this please

**Comment by rhvgoyal** [comment]
@rh-atomic-bot try

**Comment by rhvgoyal** [comment]
@rh-atomic-bot r+
